File: Recursion/lecture 8/Fraz_recursion_8.py
def perm(pos, str, ans):
    if pos>=len(str):
        ans.append(str)
    i = pos
    while i<len(str):
        str[i],str[pos]=str[pos],str[i]
        perm(pos+1,str,ans)
        str[i],str[pos]=str[pos],str[i] #backtracking
        i += 1

def generatePermutations(str):
    ans = []
    perm(0, str, ans)
    ans.sort()
    return ans

File: Recursion/lecture 8/Fraz_recursion_8.cpp
#include<bits/stdc++.h>
void perm(int pos,string &str,vector<string> &ans)
{
    if(pos>=str.size()){
        ans.push_back(str);
    }
    for(int i=pos;i<str.size();i++)
    {
        swap(str[i],str[pos]);
        perm(pos+1,str,ans);
        swap(str[i],str[pos]);//backtracking
    }
}

vector<string> generatePermutations(string &str)
{
    vector<string> ans;
    perm(0,str,ans);
    sort(ans.begin(),ans.end());
    return ans;
}